Memprof support for unmarshalled data

jhjourdan opened the pull request:
This PR implements memprof tracking for unmarshaled data.

This is performed by traversing the unmarshaled section after unmarshalling is complete. The reasons for not calling the memprof tracking function during unmarshaling are:
- The current code is less invasive in the intern.c code.
- In the current code, there is no performance penalty when memprof is disabled (while some code would have to be executed at every unmarshalled block otherwise)
- Benchmarking suggests that even when memprof is enabled, re-traversing the unmarshaled section is faster. Perhaps the reason in that the hot part of the re-traversing algorithm is a very tight loop.

In addition to this, this PR implements memprof's own root tracking mechanism for the postponed queue. Calling the `caml_xxx_global_root` functions turned out to be rather slow.

TODO:
- [x] Write tests

jhjourdan committed:
Memprof: stop using C global roots, and use our own root system.

Registering and unregistering global C roots was time consuming.

jhjourdan committed:
Memprof tracking of interned data.

gasche commented:
This stayed on my "pile of things to look at urgently" for about a week now and I haven't had the time to look, nor will I have in the closure future. Sorry for the non-progress report. Writing tests definitely sounds like a good idea to me, though.

stedolan commented:
It's also been on my pile, along with responding to your comments on #8731! I'll have time to do both on Monday, but not before.

jhjourdan commented:
No worries, I am also rather busy these days.

stedolan commented on runtime/memprof.c:
```diff
@@ -289,44 +285,38 @@ static void purge_postponed_queue(void)
+    uintnat
+      sz1 = postponed_queue_end - new_hd,
+      sz2 = new_hd - postponed_queue,
+      szof = sizeof(struct postponed_block);
```
Introducing `szof` for `sizeof (struct postponed_block)` makes this harder to read.

stedolan commented on runtime/memprof.c:
```diff
@@ -289,44 +285,38 @@ static void purge_postponed_queue(void)
        states). */
-    uintnat sz = 4 * (postponed_queue_end - postponed_queue);
+    uintnat
+      sz1 = postponed_queue_end - new_hd,
```
Is this right? It seems to include the uninitialised empty slot at `postponed_hd` in `sz1`, which gets copied below.

stedolan commented on runtime/memprof.c:
```diff
@@ -289,44 +285,38 @@ static void purge_postponed_queue(void)
-  new_hd = postponed_hd + 1;
-  if (new_hd == postponed_queue_end) new_hd = postponed_queue;
+  new_hd = postponed_next(postponed_hd);
   if (new_hd == postponed_tl) {
```
What's the motivation for reimplementing this resize code? The new version's harder to read: the old one with the while loop was obviously correct, and I'm not sure whether the new version is. Is it for performance? How often do you expect this to run?

jhjourdan replied:
I had the feeling that the `memcpy` code was more natural (eh, we are just moving some data, after all!), and yes, performance is better.

But I agree the end result is not so good. I reverted the initial code.

stedolan commented on runtime/memprof.c:
```diff
@@ -265,17 +265,13 @@ static struct postponed_block {
-  caml_remove_global_root(&postponed_tl->callstack);
-  postponed_tl++;
-  if (postponed_tl == postponed_queue_end) postponed_tl = postponed_queue;
+  if (++p == postponed_queue_end) return postponed_queue; else return p;
```
This is a bit hard to read.

stedolan commented on runtime/intern.c:
```diff
@@ -675,11 +677,32 @@ static void intern_add_to_heap(mlsize_t whsize)
     caml_allocated_words +=
       Wsize_bsize ((char *) intern_dest - intern_extra_block);
-    caml_add_to_heap(intern_extra_block);
+    if(caml_add_to_heap(intern_extra_block) != 0) {
```
This OOM check is new. Was this just wrong before?

jhjourdan replied:
Indeed.

jhjourdan replied:
Do you think this deserves its own PR? This bug is quite difficult to trigger...

stedolan replied:
No, I think it's fine as part of this one.

stedolan reviewed (changes requested):
Looks good! (This is a much smaller patch than I'd thought: I didn't realise that the regenerated `.depend` made up most of the changed lines)

I'm not completely sold on the mechanism for sharing stacks between samples within a single unmarshalling block. Making statmemprof go faster on unmarshalling allocations than on other allocations doesn't seem very important to me, because unmarshalling allocations are an uncommon case. (I'll do some benchmarking of this when I'm fixing up #8731).

jhjourdan commented:
The flambda Travis is doing a timeout. Looking at the log, it does not seem to be an infinite loop. Also, this timeout seems reproducible, since it has happened for each of the pushes of this PR.

Usually the Travis CI for flambda lasts ~30min, and the timeout is set at 50min. So, it seems that this PR created a massive slowdown. I'm investigating.

jhjourdan commented:
The reason of the slowdown seems to be what I described in #8893. This is fixed.

I wrote some tests, rebased, squashed small commits and included @stedolan comments. As far as I am concerned, this PR is ready to merge.

stedolan commented:
This looks good to me! Good to merge once CI passes (which will need some `make depend` & merge conflict fixing)

jhjourdan commented:
The conflicts are now solved, so we are just waiting for CI to succeed before merging.

jhjourdan commented:
@gasche : could you merge this? @stedolan approved it, and I'm afraid of conflicts if we wait he comes back from vacations.

gasche commented on runtime/intern.c:
```diff
@@ -776,16 +799,16 @@ static value caml_input_val_core(struct channel *chan, int outside_heap)
     intern_block = 0;
+    /* Free everything */
+    intern_cleanup();
+    return caml_check_urgent_gc(res);
```
Note: I think it would be nicer to have the same level of abstraction in the success and failure cases, with two `intern_end_{success,failure}` functions .
